pat/ReferenceTestV2/consumer/step_4.py

```diff
@@ -0,0 +1,436 @@
+    _verify_state_updates_in_time(
+        mdib=mdib,
+        step=f'{__STEP__}c',
+        node_types=(pm_qnames.AlertConditionState, pm_qnames.LimitAlertConditionState),
+        updates_required_count=5,
+        timeout=30.0,
+    )
+
+
+def test_4d(mdib: ConsumerMdib):
+    """The Reference Provider produces at least 5 alert signal updates in 30 seconds."""
+    _verify_state_updates_in_time(
+        mdib=mdib,
+        step=f'{__STEP__}d',
+        node_types=(pm_qnames.AlertConditionState, pm_qnames.LimitAlertConditionState),
+        updates_required_count=5,
+        timeout=30.0,
+    )
+
+
+def _on_alert_system_update(  # noqa: PLR0913
+    step: str,
+    descriptor_handle_to_observe: str,
+    first_update: threading.Event,
+    second_update: threading.Event,
+    last_self_check_count: dict[str, int | None],
+    alerts_by_handle: dict,
+):
+    if descriptor_handle_to_observe in alerts_by_handle:
+        state: statecontainers.AlertSystemStateContainer = alerts_by_handle[descriptor_handle_to_observe]
+        if state.SelfCheckCount is None:
+            # TODO: what to do if SelfCheckCount is None?  # noqa: FIX002, TD002, TD003
+            logger.warning(
+                'AlertSystemStateContainer "%s" has no SelfCheckCount',
+                descriptor_handle_to_observe,
+                extra={'step': step},
+            )
+            return
+        logger.debug(
+            'AlertSystemStateContainer "%s" has SelfCheckCount "%d"',
+            descriptor_handle_to_observe,
+            state.SelfCheckCount,
+            extra={'step': step},
+        )
+        previous_count = last_self_check_count.get('value')
+        if previous_count is None:
+            last_self_check_count['value'] = state.SelfCheckCount
+            first_update.set()
+            return
+        # ignore notifications keeping the counter constant (e.g. alert-condition churn) so timing uses real self checks
+        if state.SelfCheckCount != previous_count:
+            last_self_check_count['value'] = state.SelfCheckCount
+            if not first_update.is_set():
+                first_update.set()
+            else:
+                second_update.set()
+
+
+def test_4e(mdib: ConsumerMdib):
+    """The Reference Provider provides alert system self checks in accordance to the periodicity defined in the MDIB (at least every 10 seconds)."""  # noqa: E501, W505
+    step = f'{__STEP__}e'
+    max_self_check_period = 10
+    alert_systems = mdib.descriptions.NODETYPE.get(pm_qnames.AlertSystemDescriptor, [])
+    if not alert_systems:
+        result_collector.ResultCollector.log_failure(
+            step=step,
+            message='The reference provider does not provide an AlertSystemDescriptor.',
+        )
+        return
+    for alert_system in alert_systems:
+        if alert_system.SelfCheckPeriod is None:
```

Comment:
if there is no AlertSystem provided defining a SelfCheckPeriod nothing will be verified. I think if there is no such AlertSystem present a failure should be generated.

Reply:
done